File: practise/algs/sorting_algs/quick.py
from .swap import swap_elements

def partition(arr, low, high):
    i = low - 1
    pivot = arr[high]

    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            swap_elements(arr, j, i)

    swap_elements(arr, high, i + 1)
    return i + 1

def quick_sort(arr, low, high):
    if low < high:
        pivot = partition(arr, low, high)
        quick_sort(arr, low, pivot - 1)
        quick_sort(arr, pivot + 1, high)

def quick_sort2(arr, left, right):
    mid = (left + right) // 2
    pivot = arr[mid]

    swap_elements(arr, mid, right)

    i = left
    j = right - 1
    while 1:
        while arr[i] < pivot:
            i += 1
        while arr[j] > pivot:
            j -= 1

        if i >= j:
            break
        swap_elements(arr, i, j)
        
    swap_elements(arr, i, right)

    if i < right:
        quick_sort2(arr, i + 1, right)
    if j > left:
        quick_sort2(arr, left, j)
    

File: practicec/quick.c
#include <stdio.h>

void swap(int* arr, int i, int j);
void quickSort(int* arr, int low, int high);
int partition(int* arr, int low, int high);

int main()
{
    int arr[] = {8, 5, 9, 2, 6, 1, 3};   
    int n = sizeof(arr) / sizeof(arr[0]);
    quickSort(arr, 0, n - 1);
    int i;
    for(i = 0; i < n; i++)
    {
        printf("%d ", arr[i]);
    }
    printf("\n");
    return 0;
}

void swap(int* arr, int i, int j)
{
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

int partition(int* arr, int low, int high)
{
    int mid, i, j, pivot;
    mid = (high + low) / 2;
    pivot = arr[mid];

    swap(arr, mid, high);

    i = low - 1;
    for(j = low; j < high; j++)
    {
        if(arr[j] < pivot)
        {
            i++;
            swap(arr, i, j);
        }
    }
    swap(arr, i + 1, high);
    return i + 1;
}

void quickSort(int* arr, int low, int high)
{
    if(low < high)
    {
        int pivot = partition(arr, low, high);
        quickSort(arr, low, pivot - 1);
        quickSort(arr, pivot + 1, high);
    }
}